Mathlib/Analysis/Meromorphic/Divisor/MeromorphicFunction.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2025 Stefan Kebekus. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stefan Kebekus
+-/
+
+import Mathlib.Analysis.Meromorphic.Order
+import Mathlib.Analysis.Meromorphic.Divisor.Basic
+
+/-!
+# The Divisor of a Meromorphic Function
+
+This file defines the divisor of a meromorphic function and proves the most
+basic lemmas about those divisors.
+
+## TODO
+
+- Remove the assumption `CompleteSpace E`.
+- Compatibility with restriction of divisors/functions
+- Non-negativity of the divisor for an analytic function
+- Behavior under addition of functions
+- Congruence lemmas for `codiscreteWithin`
+-/
+
+open Classical
+
+variable
+  {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {U : Set 𝕜}
+  {z : 𝕜}
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+
+namespace MeromorphicOn
+
+/-!
+## Definition of the Divisor
+-/
+
+/-- The divisor of a meromorphic function `f`, mapping a point `z` to the order
+  of `f` at `z`, and to zero if the order is infinite. -/
+noncomputable def divisorOn {f : 𝕜 → E} (hf : MeromorphicOn f U) :
```

Comment:
I advise to make `f` explicit as well.
Lean has the tendency to hide complicated proof terms as ellipsis `...` which can become quite unhelpful if you are trying to figure out which function you are taking the divisor of.

In fact, you could consider defining the divisor for *all* functions, by defining it to be `0` for ones that are not meromorphic. If you decide to do that `hf` will of course remain an argument of most theorems below.
Also: in that case `U` should probably become explicit.

Reply:
I perfectly understand what you are saying [constantly moving my mouse over the ellipsis to see what function we are discussing], but I need time to form an opinion.

This is a terrible trade-off. On the one hand, ellipsis are a pain. On the other hand, the notation is extremely compact and intuitive as it is. Writing `hf.divisorOn` is so much better than `MeromorphicOn.divisorOn f hf`. I need to experiment a little to see what variant will work for me.

Reply:
The suggestion defining divisors for all functions has its charm, but might later lead to conflicts when one later tries to attach a divisor to functions in other contexts. I am thinking about divisors encoding growth rates of  (sub)harmonic functions, level sets of Lelong numbers, multiplier ideals in the theory of PDEs, …

Reply:
Those other contexts will have their own definition, in its own namespace. So I don't think there will be a conflict.

And with `open MeromorphicOn` you would end up writing `divisorOn f hf`, which is maybe not so bad.

Reply:
Ok, that got me convinced. Implemented according to your suggestion, only writing `divisor` instead of `divisorOn` because I feel if wouldn't be quite obvious what the `On` refers to.

Reply:
Returning to this, I think Mario points out a good argument in https://leanprover.zulipchat.com/#narrow/channel/113489-new-members/topic/Square.20root.20of.20-1.20as.20a.20Real/near/505513202 to totalize `divisor` and only take `f` (and `U`) as arguments, and drop `hf`.

Reply:
Will implement.

Reply:
Implemented. @jcommelin Now how do you like that?